File: product_sum.rb
def product_sum(array)
  product_summer(array, 1)
end

def product_summer(array, depth)
  sum = 0

  array.each do |item|
    sum += if item.instance_of?(Integer)
      item
    else
      product_summer(item, depth + 1)
    end
  end

  sum * depth
end

puts product_sum([1, 2, 3]) #=> 6
puts product_sum([1, 2, [3]]) #=> 9

File: product_sum.py
# Tip: You can use the type(element) function to check whether an item
# is a list or an integer.
def productSum(array):
    return productSummer(array, 1)

def productSummer(array, depth):
    sum = 0
    for item in array:
        if type(item) is int:
            sum += item
        else:
            sum += productSummer(item, depth + 1)

    return sum * depth


print(productSum([1, 2, 3])) #=> 6
print(productSum([1, 2, [3]])) #=> 12
